iot-e2e-tests/common/src/test/java/tests/integration/com/microsoft/azure/sdk/iot/helpers/rules/ThrottleResistantTestRule.java

```diff
@@ -0,0 +1,55 @@
+package tests.integration.com.microsoft.azure.sdk.iot.helpers.rules;
+
+import com.microsoft.azure.sdk.iot.service.exceptions.IotHubTooManyRequestsException;
+import lombok.extern.slf4j.Slf4j;
+import org.junit.Assume;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+import tests.integration.com.microsoft.azure.sdk.iot.helpers.IntegrationTest;
+import tests.integration.com.microsoft.azure.sdk.iot.helpers.annotations.StandardTierHubOnlyTest;
+
+/**
+ * On lower tier IoT Hubs, this test project often hits throttling limits. Rather than add retry logic to every call
+ * to the hub, this rule will catch these throttling errors and re-run the test until it passes or fails for non-throttling
+ * reasons. It has a configurable delay between each re-run to give IoT Hub time to free up some throughput
+ */
+@Slf4j
+public class ThrottleResistantTestRule implements TestRule
+{
+    final static int THROTTLING_RETRY_DELAY_MILLISECONDS = 10 * 1000;
+
+    @Override
+    public Statement apply(Statement base, Description description) {
+        return new ThrottleResistantStatement(base, description);
+    }
+
+    private class ThrottleResistantStatement extends Statement {
+
+        private final Statement base;
+        private final Description description;
+
+        public ThrottleResistantStatement(Statement base, Description description) {
+            this.base = base;
+            this.description = description;
+        }
+
+        @Override
+        public void evaluate() throws Throwable {
+            boolean ranWithoutThottling = false;
+            while (!ranWithoutThottling)
+            {
+                try
+                {
+                    base.evaluate();
+                    ranWithoutThottling = true;
+                }
+                catch (IotHubTooManyRequestsException e)
```

Comment:
This exception gets thrown by any HTTP request (registryManager.addDevice, deviceMethod.invokeMethod, etc.)